File: ios/Classes/FlutterflappytoolsPlugin.m
#import "FlutterflappytoolsPlugin.h"
#import "AudioToolbox/AudioToolbox.h"
#import <UserNotifications/UserNotifications.h>
#import <AVFoundation/AVFoundation.h>
#import <MapKit/MapKit.h>
#import <Photos/Photos.h>



@implementation FlutterflappytoolsPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    
    FlutterflappytoolsPlugin* instance = [[FlutterflappytoolsPlugin alloc] init];
    //create method channel
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"flutterflappytools"
                                     binaryMessenger:[registrar messenger]];
    [registrar addMethodCallDelegate:instance channel:channel];
    
    
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    if([@"getPathSize" isEqualToString:call.method]){
        NSString* path=(NSString*)call.arguments[@"path"];
        NSString* type=(NSString*)call.arguments[@"type"];
        long long retSize=[self caculateSize:path];
        NSString* retStr=@"";
        //B
        if(type.intValue==1){
            retStr=[NSString stringWithFormat:@"%.2f",(float)retSize];
        }
        //KB
        else if(type.intValue==2){
            retStr=[NSString stringWithFormat:@"%.2f",(float)retSize/1024];
        }
        //MB
        else if(type.intValue==3){
            retStr=[NSString stringWithFormat:@"%.2f",(float)retSize/1048576];
        }
        //GB
        else if(type.intValue==4){
            retStr=[NSString stringWithFormat:@"%.2f",(float)retSize/1073741824];
        }
        result(retStr);
    }
    //clear path
    else if([@"clearPath" isEqualToString:call.method]){
        NSString* path=(NSString*)call.arguments[@"path"];
        NSFileManager *fileManager = [NSFileManager defaultManager];
        [fileManager removeItemAtPath:path error:nil];
        result(@"1");
    }
    //get brightness
    else  if([@"getBrightness" isEqualToString:call.method]){
        CGFloat brightness=[UIScreen mainScreen].brightness;
        result([NSString stringWithFormat:@"%.2f",brightness]);
    }
    //get device current locale
    else  if([@"getDeviceLocal" isEqualToString:call.method]){
        UIDevice* currentDevice = [UIDevice currentDevice];
        NSArray*  languageArray = [NSLocale preferredLanguages];
        if(languageArray.count==0){
            result(@"en-US");
            return;
        }
        NSString* language = [languageArray objectAtIndex:0];
        result(language);
    }
    //setBrightness
    else if([@"setBrightness" isEqualToString:call.method]){
        NSString* brightness=(NSString*)call.arguments[@"brightness"];
        CGFloat fla=brightness.doubleValue;
        [[UIScreen mainScreen] setBrightness:fla];
        result(@"1");
    }
    //getBatteryLevel
    else if([@"getBatteryLevel" isEqualToString:call.method]){
        [UIDevice currentDevice].batteryMonitoringEnabled = YES;
        double deviceLevel = [UIDevice currentDevice].batteryLevel;
        result([NSString stringWithFormat:@"%f",deviceLevel]);
    }
    //getBatteryCharge
    else if([@"getBatteryCharge" isEqualToString:call.method]){
        if([UIDevice currentDevice].batteryState==UIDeviceBatteryStateCharging){
            result(@"1");
        }else{
            result(@"0");
        }
    }
    //getBatteryCharge
    else if([@"checkPermission" isEqualToString:call.method]){
        NSInteger type=((NSString*)call.arguments[@"type"]).intValue;
        if(type==0){
            [[UNUserNotificationCenter currentNotificationCenter]
             requestAuthorizationWithOptions:(UNAuthorizationOptionAlert + UNAuthorizationOptionSound + UNAuthorizationOptionBadge)
             completionHandler:^(BOOL granted, NSError * _Nullable error) {
                if(granted){
                    result(@"1");
                } else {
                    result(@"0");
                }
            }];
            return;
        }
        if(type==1){
            [AVCaptureDevice requestAccessForMediaType:AVMediaTypeVideo completionHandler:^(BOOL granted) {
                if(granted){
                    result(@"1");
                } else {
                    result(@"0");
                }
            }];
            return;
        }
        if(type==2){
            [PHPhotoLibrary requestAuthorization:^(PHAuthorizationStatus status) {
                switch (status) {
                    case PHAuthorizationStatusLimited:
                    case PHAuthorizationStatusAuthorized:
                        result(@"1");
                        break;
                    case PHAuthorizationStatusDenied:
                        result(@"0");
                        break;
                    case PHAuthorizationStatusNotDetermined:
                        result(@"0");
                        break;
                    case PHAuthorizationStatusRestricted:
                        result(@"0");
                        break;
                }
            }];
            return;
        }
        
    }
    //setSceenSteadyLight
    else if([@"setScreenSteadyLight" isEqualToString:call.method]){
        NSString* state=(NSString*)call.arguments[@"state"];
        if([state isEqualToString:@"1"]){
            [[UIApplication sharedApplication] setIdleTimerDisabled:YES];
        }else{
            [[UIApplication sharedApplication] setIdleTimerDisabled:NO];
        }
        result(@"1");
    }
    //shake
    else  if([@"shake" isEqualToString:call.method]){
        AudioServicesPlaySystemSound(kSystemSoundID_Vibrate);
        result(@"1");
    }
    //cookie
    else  if([@"addManagerCookie" isEqualToString:call.method]){
        NSString* cookie=(NSString*)call.arguments[@"cookie"];
        //URL
        NSString* url=(NSString*)call.arguments[@"url"];
        //array
        NSArray *aArray = [cookie componentsSeparatedByString:@"="];
        //alloc
        NSDictionary* properties = [[NSMutableDictionary alloc] init];
        [properties setValue:aArray[1] forKey:NSHTTPCookieValue];
        [properties setValue:aArray[0] forKey:NSHTTPCookieName];
        [properties setValue:[[NSURL URLWithString:url] host] forKey:NSHTTPCookieDomain];
        [properties setValue:@"/" forKey:NSHTTPCookiePath];
        [properties setValue:@"0" forKey:NSHTTPCookieVersion];
        [properties setValue:[NSDate dateWithTimeIntervalSinceNow:3600*24*30*12] forKey:NSHTTPCookieExpires];
        NSHTTPCookie* cook = [[NSHTTPCookie alloc] initWithProperties:properties];
        [[NSHTTPCookieStorage sharedHTTPCookieStorage] setCookie:cook];
        result(@"1");
    }
    else if([@"share" isEqualToString:call.method]){
        NSString* share=(NSString*)call.arguments[@"share"];
        NSArray * activityItems = [[NSArray alloc] initWithObjects:share, nil];
        UIActivityViewController * activityVC = [[UIActivityViewController alloc]initWithActivityItems:activityItems applicationActivities:nil];
        UIActivityViewControllerCompletionWithItemsHandler myBlock = ^(UIActivityType activityType, BOOL completed, NSArray *returnedItems, NSError *activityError) {
            NSLog(@"%@",activityType);
            if (completed) {
                NSLog(@"Share success");
            } else {
                NSLog(@"Share failure");
            }
            [activityVC dismissViewControllerAnimated:YES completion:nil];
        };
        activityVC.completionWithItemsHandler = myBlock;
        UIViewController *topController = [self _topViewController:[[UIApplication sharedApplication].keyWindow rootViewController]];
        
        [topController presentViewController:activityVC animated:YES completion:nil];
        
        result(@"true");
    }
    else if([@"jumpToUrl" isEqualToString:call.method]){
        NSString* url=(NSString*)call.arguments[@"url"];
        [self openScheme:url];
        result(@"true");
    }
    else if([@"jumpToScheme" isEqualToString:call.method]){
        NSString* url=(NSString*)call.arguments[@"url"];
        [self openScheme:url];
        result(@"true");
    }
    //is map install
    else if([@"isMapInstalled" isEqualToString:call.method]){
        //type
        NSInteger type=((NSString*)call.arguments[@"type"]).intValue;
        //apple
        if(type==0){
            result(@"1");
        }
        //amap
        else if(type==1){
            if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"iosamap://"]]) {
                result(@"1");
            }else{
                result(@"0");
            }
        }
        //baidu
        else if(type==2){
            if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"baidumap://"]]) {
                result(@"1");
            }else{
                result(@"0");
            }
        }
        //tencent
        else if(type==3){
            if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"qqmap://"]]) {
                result(@"1");
            }else{
                result(@"0");
            }
        }else{
            result(@"0");
        }
    }
    //jump to map
    else if([@"jumpToMap" isEqualToString:call.method]){
        //map type
        NSInteger type=((NSString*)call.arguments[@"type"]).intValue;
        NSString* lat=(NSString*)call.arguments[@"lat"];
        NSString* lng=(NSString*)call.arguments[@"lng"];
        NSString* title=(NSString*)call.arguments[@"title"];
        //apple
        if(type==0){
            CLLocationCoordinate2D loc = CLLocationCoordinate2DMake([lat floatValue],
                                                                    [lng floatValue]);
            MKMapItem *currentLocation = [MKMapItem mapItemForCurrentLocation];
            MKMapItem *toLocation = [[MKMapItem alloc] initWithPlacemark:[[MKPlacemark alloc] initWithCoordinate:loc addressDictionary:nil]];
            [MKMapItem openMapsWithItems:@[currentLocation, toLocation]
                           launchOptions:@{MKLaunchOptionsDirectionsModeKey: MKLaunchOptionsDirectionsModeDriving,
                                           MKLaunchOptionsShowsTrafficKey: [NSNumber numberWithBool:YES]}];
            result(@"1");
        }
        //amap
        else if(type==1){
            // check amap
            if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"iosamap://"]]) {
                
                NSString *urlString = [[NSString stringWithFormat:@"iosamap://navi?sourceApplication=%@&poiname=%@&backScheme=%@&lat=%f&lon=%f&dev=0&style=2",
                                        @"Navigation",
                                        title,
                                        @"",
                                        [lat floatValue],
                                        [lng floatValue]] stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]];
                if (@available(iOS 10.0, *)) {
                    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:urlString] options:@{} completionHandler:nil];
                }else{
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
                    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:urlString]];
#pragma clang diagnostic pop
                }
                result(@"1");
            }else{
                result(@"0");
            }
        }
        //baidumap
        else if(type==2){
            // check baidumap
            if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"baidumap://"]]) {
                
                NSString *urlString = [[NSString stringWithFormat:@"baidumap://map/direction?origin={{Location}}&destination=latlng:%f,%f|name=%@&mode=driving&coord_type=gcj02",[lat floatValue],[lng floatValue],title] stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]];
                if (@available(iOS 10.0, *)) {
                    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:urlString] options:@{} completionHandler:nil];
                }else{
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
                    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:urlString]];
#pragma clang diagnostic pop
                }
                result(@"1");
            }else{
                result(@"0");
            }
        }
        //tencent map
        else if(type==3){
            // tencent map
            if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"qqmap://"]]) {
                NSString *urlString = [[NSString stringWithFormat:@"qqmap://map/routeplan?type=drive&fromcoord=CurrentLocation&tocoord=%@,%@&referer=IXHBZ-QIZE4-ZQ6UP-DJYEO-HC2K2-EZBXJ",
                                        lat,
                                        lng] stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]];
                if (@available(iOS 10.0, *)) {
                    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:urlString] options:@{} completionHandler:nil];
                }else{
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
                    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:urlString]];
#pragma clang diagnostic pop
                }
                result(@"1");
            }else{
                result(@"0");
            }
        }else{
            result(@"0");
        }
    }
    else {
        result(FlutterMethodNotImplemented);
    }
}



//goto scheme
- (void)openScheme:(NSString *)scheme {
    NSString* encodedString =  [scheme stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet URLQueryAllowedCharacterSet]];
    UIApplication *application = [UIApplication sharedApplication];
    NSURL *URL = [NSURL URLWithString:encodedString];
    if (@available(iOS 10.0, *)) {
        [application openURL:URL options:@{}
           completionHandler:^(BOOL success) {
            NSLog(@"Open %@: %d",scheme,success);
        }];
    } else {
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
        [application openURL:URL];
#pragma clang diagnostic pop
    }
}

//get top controller
- (UIViewController *)_topViewController:(UIViewController *)vc {
    if ([vc isKindOfClass:[UINavigationController class]]) {
        return [self _topViewController:[(UINavigationController *)vc topViewController]];
    } else if ([vc isKindOfClass:[UITabBarController class]]) {
        return [self _topViewController:[(UITabBarController *)vc selectedViewController]];
    } else {
        return vc;
    }
    return nil;
}

//caculateSize
-(long long)caculateSize:(NSString*) filePath{
    if([self isDirectory:filePath]){
        return [self folderSizeAtPath:filePath];
    }else{
        return [self fileSizeAtPath:filePath];
    }
}

//check isDirectory
- (BOOL)isDirectory:(NSString *)filePath
{
    BOOL isDirectory = NO;
    [[NSFileManager defaultManager] fileExistsAtPath:filePath isDirectory:&isDirectory];
    return isDirectory;
}

//path size
- (long long) fileSizeAtPath:(NSString*) filePath{
    NSFileManager* manager = [NSFileManager defaultManager];
    if ([manager fileExistsAtPath:filePath]){
        return [[manager attributesOfItemAtPath:filePath error:nil] fileSize];
    }
    return 0;
}

//folder size
- (long long) folderSizeAtPath:(NSString*) folderPath{
    NSFileManager* manager = [NSFileManager defaultManager];
    if (![manager fileExistsAtPath:folderPath]) return 0;
    NSEnumerator *childFilesEnumerator = [[manager subpathsAtPath:folderPath] objectEnumerator];
    NSString* fileName;
    long long folderSize = 0;
    while ((fileName = [childFilesEnumerator nextObject]) != nil){
        NSString* fileAbsolutePath = [folderPath stringByAppendingPathComponent:fileName];
        folderSize += [self fileSizeAtPath:fileAbsolutePath];
    }
    return folderSize;
}

@end

File: ios/Classes/FlutterFlappyToolsPlugin.swift
import Flutter
import UIKit
import AudioToolbox
import AVFoundation
import Photos

public class FlutterFlappyToolsPlugin: NSObject, FlutterPlugin {
    // 注册插件并创建方法通道
    public static func register(with registrar: FlutterPluginRegistrar) {
        // 创建方法通道
        let channel = FlutterMethodChannel(name: "flutter_flappy_tools", binaryMessenger: registrar.messenger())
        let instance = FlutterFlappyToolsPlugin()
        // 将插件实例注册为方法通道的委托
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    // 处理来自 Flutter 的方法调用
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPathSize":
            // 获取路径大小
            if let args = call.arguments as? [String: Any],
               let path = args["path"] as? String,
               let type = args["type"] as? Int {
                let size = calculateSize(path: path)
                let formattedSize = formatSize(size: size, type: type)
                result(formattedSize)
            } else {
                // 参数错误时返回错误
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for getPathSize", details: nil))
            }
        case "clearPath":
            // 清除指定路径的文件
            if let args = call.arguments as? [String: Any],
               let path = args["path"] as? String {
                clearPath(path: path)
                result("1")
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for clearPath", details: nil))
            }
        case "getBrightness":
            // 获取屏幕亮度
            let brightness = UIScreen.main.brightness
            result(String(format: "%.2f", brightness))
        case "getDeviceLocal":
            // 获取设备的当前语言设置
            let language = Locale.preferredLanguages.first ?? "en-US"
            result(language)
        case "setBrightness":
            // 设置屏幕亮度
            if let args = call.arguments as? [String: Any],
               let brightness = args["brightness"] as? String,
               let value = Double(brightness) {
                UIScreen.main.brightness = CGFloat(value)
                result("1")
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for setBrightness", details: nil))
            }
        case "getBatteryLevel":
            // 获取设备电池电量
            UIDevice.current.isBatteryMonitoringEnabled = true
            let batteryLevel = UIDevice.current.batteryLevel
            result(String(format: "%.2f", batteryLevel))
        case "getBatteryCharge":
            // 检查设备是否正在充电
            UIDevice.current.isBatteryMonitoringEnabled = true
            let isCharging = UIDevice.current.batteryState == .charging
            result(isCharging ? "1" : "0")
        case "setScreenSteadyLight":
            // 设置屏幕常亮状态
            if let args = call.arguments as? [String: Any],
               let state = args["state"] as? String {
                UIApplication.shared.isIdleTimerDisabled = (state == "1")
                result("1")
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for setScreenSteadyLight", details: nil))
            }
        case "shake":
            // 触发设备震动
            AudioServicesPlaySystemSound(kSystemSoundID_Vibrate)
            result("1")
        case "addManagerCookie":
            // 添加 Cookie
            if let args = call.arguments as? [String: Any],
               let cookie = args["cookie"] as? String,
               let url = args["url"] as? String {
                addCookie(cookie: cookie, url: url)
                result("1")
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for addManagerCookie", details: nil))
            }
        case "share":
            // 调用系统分享功能
            if let args = call.arguments as? [String: Any],
               let text : String = args["share"] as? String {
                shareText(text: text, result: result)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for share", details: nil))
            }
        case "jumpToUrl", "jumpToScheme":
            // 打开指定的 URL 或 Scheme
            if let args = call.arguments as? [String: Any],
               let url = args["url"] as? String {
                openScheme(url: url)
                result("true")
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for jumpToUrl/jumpToScheme", details: nil))
            }
        default:
            // 未实现的方法
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - Helper Methods
    
    // 计算路径的大小（包括文件和文件夹）
    private func calculateSize(path: String) -> Int64 {
        let fileManager = FileManager.default
        var isDirectory: ObjCBool = false
        if fileManager.fileExists(atPath: path, isDirectory: &isDirectory) {
            if isDirectory.boolValue {
                return folderSize(atPath: path)
            } else {
                return fileSize(atPath: path)
            }
        }
        return 0
    }
    
    // 获取文件的大小
    private func fileSize(atPath path: String) -> Int64 {
        let fileManager = FileManager.default
        if let attributes = try? fileManager.attributesOfItem(atPath: path),
           let fileSize = attributes[.size] as? Int64 {
            return fileSize
        }
        return 0
    }
    
    // 获取文件夹的大小
    private func folderSize(atPath path: String) -> Int64 {
        let fileManager = FileManager.default
        guard let enumerator = fileManager.enumerator(atPath: path) else { return 0 }
        var totalSize: Int64 = 0
        for case let file as String in enumerator {
            let filePath = (path as NSString).appendingPathComponent(file)
            totalSize += fileSize(atPath: filePath)
        }
        return totalSize
    }
    
    // 格式化文件大小为指定单位
    private func formatSize(size: Int64, type: Int) -> String {
        switch type {
        case 1: return String(format: "%.2f", Double(size))
        case 2: return String(format: "%.2f", Double(size) / 1024)
        case 3: return String(format: "%.2f", Double(size) / 1048576)
        case 4: return String(format: "%.2f", Double(size) / 1073741824)
        default: return "0"
        }
    }
    
    // 清除指定路径的文件
    private func clearPath(path: String) {
        let fileManager = FileManager.default
        try? fileManager.removeItem(atPath: path)
    }
    
    // 添加 Cookie 到指定的 URL
    private func addCookie(cookie: String, url: String) {
        let components = cookie.split(separator: "=")
        guard components.count == 2 else { return }
        let name = String(components[0])
        let value = String(components[1])
        let properties: [HTTPCookiePropertyKey: Any] = [
            .name: name,
            .value: value,
            .domain: URL(string: url)?.host ?? "",
            .path: "/",
            .version: "0",
            .expires: Date(timeIntervalSinceNow: 3600 * 24 * 30 * 12)
        ]
        if let cookie = HTTPCookie(properties: properties) {
            HTTPCookieStorage.shared.setCookie(cookie)
        }
    }
    
    // 调用系统分享功能
    private func shareText(text: String, result: @escaping FlutterResult) {
        let activityVC = UIActivityViewController(activityItems: [text], applicationActivities: nil)
        if let topController = UIApplication.shared.keyWindow?.rootViewController {
            topController.present(activityVC, animated: true, completion: nil)
        }
        result("true")
    }
    
    // 打开指定的 URL 或 Scheme
    private func openScheme(url: String) {
        if let encodedURL = url.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
           let url = URL(string: encodedURL) {
            UIApplication.shared.open(url, options: [:], completionHandler: nil)
        }
    }
}
